bootstrap/src/lib/cache/index.ts

```diff
@@ -97,19 +109,29 @@ export const withCache: Middleware<CacheOptions> = async (execute, options = def
     logger.debug(`Request coalescing: DEL ${key}`)
   }
 
-  const _getMaxAge = (data: AdapterRequest): any => {
-    if (!data || !data.data) return cache.options.maxAge
-    if (isNaN(data.data.maxAge as number)) return cache.options.maxAge
-    return Number(data.data.maxAge) || cache.options.maxAge
+  const _getRequestMaxAge = (data: AdapterRequest): number | undefined => {
+    if (!data || !data.data) return
+    if (isNaN(data.data.maxAge as number)) return
+    return Number(data.data.maxAge)
+  }
+  const _getDefaultMaxAge = (request: AdapterRequest): number => {
+    return rateLimit.getParticipantMaxAge(_getKey(request)) || (cache.options.maxAge as number)
   }
 
-  const _executeWithCache = async (data: AdapterRequest) => {
-    const key = _getKey(data)
+  const _executeWithCache = async (request: AdapterRequest) => {
+    const key = _getKey(request)
     const coalescingKey = _getCoalescingKey(key)
-    const maxAge = _getMaxAge(data)
     // Add successful result to cache
-    const _cacheOnSuccess = async ({ statusCode, data, result }: AdapterResponse) => {
+    const _cacheOnSuccess = async (
+      { statusCode, data, result }: AdapterResponse,
+      maxAge?: number,
+    ) => {
       if (statusCode === 200) {
+        // Heartbeat is incremented only if valid request
+        rateLimit.incrementParticipantHeartbeat(_getKey(request), data.cost)
+        if (!maxAge) {
```

Comment:
This basically gives the option to override the default maxAge calculation, useful when it comes in the request